Initial FormCommit tests

sharwell opened the pull request:
I added some notes to the review which are relevant to how this expands from testing a single form to testing many.

📝 If accepted, please do not rebase or squash this pull request during the merge.

sharwell commented:
I updated this pull request so GitHub would only consider the new commits, and am hoping it's enough to get an accurate picture for baseline code coverage of this dialog that we can build on for some other open bug fix PRs. Will review results and update in the morning.

sharwell commented on GitUI/CommandsDialogs/FormCommit.cs:
```diff
@@ -3152,6 +3153,23 @@ private void Message_Enter(object sender, EventArgs e)
         }
+
+        internal TestAccessor GetTestAccessor()
+            => new TestAccessor(this);
```
:memo: This is a common pattern we used at a former company to avoid production code "leaking" private objects. We found that consistent naming made it virtually impossible for any use of `GetTestAccessor` in production code to make it through code review.

sharwell commented on GitUI/CommandsDialogs/FormCommit.cs:
```diff
@@ -3152,6 +3153,23 @@ private void Message_Enter(object sender, EventArgs e)
+                _formCommit = formCommit;
+            }
+
+            public EditNetSpell Message => _formCommit.Message;
```
:memo: There is likely no need to use `TestAccessor` for purposes beyond exposing controls. Anything else would be considered a red flag (likely misuse or unnecessarily complex).

sharwell commented on UnitTests/GitUITests/CommandsDialogs/CommitDialog/FormCommitTests.cs:
```diff
@@ -0,0 +1,209 @@
+        {
+            _moduleTestHelper = new GitModuleTestHelper();
+
+            using (var repository = new LibGit2Sharp.Repository(_moduleTestHelper.Module.WorkingDir))
```
💭 In the long run, we would likely see benefits from creating a test state shared across a wider variety of tests. This object can quickly create a repository with a known form, and reset its state between tests so it's easy to write a test that relies on a current commit with particular characteristics. The use of LibGit2Sharp has two advantages:

1. It's a simple and fast API for constructing a repository with a known form
2. It's decoupled from GitExtensions action-tracking features, so it's easy to create a repository during testing without otherwise influencing the GitExtensions behavior

sharwell commented on UnitTests/GitUITests/CommandsDialogs/CommitDialog/FormCommitTests.cs:
```diff
@@ -0,0 +1,209 @@
+            });
+        }
+
+        private static async Task WaitForIdleAsync()
```
💭 I would consider this a candidate for refactoring once we establish some initial patterns.

sharwell commented on UnitTests/GitUITests/CommandsDialogs/CommitDialog/FormCommitTests.cs:
```diff
@@ -0,0 +1,209 @@
+                commitKind);
+        }
+
+        private void RunFormCommitTest(Func<FormCommit, Task> testDriverAsync, CommitKind commitKind = CommitKind.Normal)
```
💭 It may be possible to further extract a shared test helper `RunFormTest`, but it didn't seem necessary for this initial implementation.

RussKie commented:
Very nice!

On 28 March 2018 at 15:07, Sam Harwell <notifications@github.com> wrote:

> I updated this pull request so GitHub would only consider the new commits,
> and am hoping it's enough to get an accurate picture for baseline code
> coverage of this dialog that we can build on for some other open bug fix
> PRs. Will review results and update in the morning.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/4717#issuecomment-376754891>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEMyXrLv_cpkFwAM5xT7MUYhxWzNDMj6ks5tiwyKgaJpZM4S8-hf>
> .
>

AArnott commented on UnitTests/GitUITests/CommandsDialogs/CommitDialog/FormCommitTests.cs:
```diff
@@ -0,0 +1,209 @@
+
+            // Queue an event to make sure we don't stall if the application was already idle
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+            await Task.Yield();
```
You know, we have this exact problem in VS where folks might schedule idle work, but idle is only executed on the transition from a non-empty queue to an empty queue. We call it "tickling idle" when you have to post a message just to get idle work to process again. 😄

RussKie commented:
Very interesting. Though we are shipping a test code into production which is undesirable...
Can we create a different build configuration and decorate these test fixtures with conditional attributes so they don't exist in the final production output?

sharwell commented:
> Very interesting. Though we are shipping a test code into production which is undesirable...

At a former employer, we ended up deciding that among known alternatives, this approach was the cleanest.

1. The accessor has `internal` visibility
2. The accessor is always named `GetTestAccessor()`, which is already easy in code review to avoid ever calling in production code but an analyzer could be added in theory
3. The contents of the accessor are small

RussKie commented:
Please resolve MC and merge it in, thanks

sharwell commented:
@RussKie I rebased this to fix the conflicts, which is the first four commits. The last three commits are new; please review this refactoring and confirm that the direction looks good.

sharwell commented:
A few corrections after rebase:

1. This change reveals a regression introduced by #4699. I added `[Ignore]` to the regression test which reveals the problem and filed #4856 to address it.
2. Due to a limitation in NUnit initialization order, it is not possible to use `ThreadHelper.JoinableTaskFactory` from a method marked with `[OneTimeSetUp]`. The commits were amended to use `[SetUp]` instead
3. The version of NuGet included in the repository is 4.1. Version 4.3 is required to use a **netstandard2.0** package from a **net461** project. I filed #4857 to fix this properly, and implemented a temporary workaround in 2ab47f2.

drewnoakes commented on UnitTests/GitUITests/CommandsDialogs/ReferenceRepository.cs:
```diff
@@ -0,0 +1,58 @@
+
+namespace GitUITests.CommandsDialogs
+{
+    public class ReferenceRepository : IDisposable
```
Was the intention for this class to have a destructor?

sharwell replied:
No

drewnoakes replied:
Why the finalizer/disposable pattern here then?

sharwell replied:
This pattern applies to any class which is not `sealed`.

drewnoakes replied:
Subclasses can override `Dispose` and still add this logic if they have unmanaged resources. Am I missing something, or is this a matter of preference?

drewnoakes commented on UnitTests/GitUITests/CommandsDialogs/CommitDialog/FormCommitTests.cs:
```diff
@@ -0,0 +1,141 @@
+
+                // Verify the message appears correctly
+                commitMessageToolStripMenuItem.ShowDropDown();
+                Assert.AreEqual("A commit message...", commitMessageToolStripMenuItem.DropDownItems[0].Text);
```
#4867 fixes the blank message bug, but you'll need to update the expected string to remove the ellipsis to make the test pass.

drewnoakes commented:
#4867 has been merged the ignored test can be reinstated.

sharwell commented:
Rebased again to account for #4867

RussKie commented:
It looks good to me, though I haven't pulled it down to run.
It looks intricate and may take me some time to get comfortable writing UI-specific tests before I can form an opinion about how useable the proposed API is. 
In either case it is a good start.
